### ISSUE TITLE
Stopped polling job due to exception: System.IO.IOException

### ISSUE THREAD

**Issue by exzizt** [googlesamples/unity-jar-resolver#141]
This exception occurs in the Unity Editor periodically. Unity version 2018.2.1f1.

> Stopped polling job due to exception: System.IO.IOException: Sharing violation on path Temp\GoogleAarExplodeCache.xml
>   at System.IO.File.Delete (System.String path) [0x00073] in <f826c2584fc94ec19a48a6576640bdc5>:0 
>   at GooglePlayServices.ResolverVer1_1.SaveAarExplodeCache () [0x00028] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.ResolverVer1_1.OnBuildSettings () [0x00125] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.PlayServicesResolver.ResolveUnsafe (System.Action`1[T] resolutionComplete, System.Boolean forceResolution) [0x00017] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.PlayServicesResolver+<ScheduleResolve>c__AnonStorey10.<>m__1B () [0x00000] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.PlayServicesResolver.ExecuteNextResolveJob () [0x00069] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.PlayServicesResolver.ScheduleResolve (System.Boolean forceResolution, System.Action`1[T] resolutionCompleteWithResult, System.Boolean isAutoResolveJob) [0x00094] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.PlayServicesResolver.AutoResolve (System.Action resolutionComplete) [0x0001c] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at GooglePlayServices.PlayServicesResolver.<ScheduleAutoResolve>m__13 () [0x0003b] in <6804ee413fd14b908a616b3fd8ae92db>:0 
>   at Google.RunOnMainThread+ScheduledJob.PollUntilExecutionTime () [0x00035] in <68420a0241b140c584225cd97e980f81>:0 
>   at Google.RunOnMainThread.ExecutePollingJobs () [0x0004a] in <68420a0241b140c584225cd97e980f81>:0 
> UnityEngine.Debug:LogError(Object)
> Google.RunOnMainThread:ExecutePollingJobs()
> Google.RunOnMainThread:ExecuteAll()
> UnityEditor.EditorApplication:Internal_CallUpdateFunctions()

**Comment by stewartmiles** [comment]
@exzizt your `Temp` directory must be read-only.  The resolver uses the directory to cache AAR / JAR state to speed up auto-resolution.  You really shouldn't be checking the `Temp` directory into source control.

**Comment by stewartmiles** [comment]
Well this does appear to be a bug on my Windows box as well.  I see the following happen after resolution:

````
Stopped polling job due to exception: System.IO.IOException: Sharing violation on path Temp\GoogleAarExplodeCache.xml
  at System.IO.File.Delete (System.String path) [0x000b2] in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.IO/File.cs:179 
  at GooglePlayServices.ResolverVer1_1.SaveAarExplodeCache () [0x00000] in <filename unknown>:0 
  at GooglePlayServices.ResolverVer1_1.OnBuildSettings () [0x00000] in <filename unknown>:0 
  at GooglePlayServices.PlayServicesResolver.ResolveUnsafe (System.Action`1 resolutionComplete, Boolean forceResolution) [0x00000] in <filename unknown>:0 
  at GooglePlayServices.PlayServicesResolver+<ScheduleResolve>c__AnonStorey3.<>m__0 () [0x00000] in <filename unknown>:0 
  at GooglePlayServices.PlayServicesResolver.ExecuteNextResolveJob () [0x00000] in <filename unknown>:0 
  at GooglePlayServices.PlayServicesResolver.ScheduleResolve (Boolean forceResolution, System.Action`1 resolutionCompleteWithResult, Boolean isAutoResolveJob) [0x00000] in <filename unknown>:0 
  at GooglePlayServices.PlayServicesResolver.AutoResolve (System.Action resolutionComplete) [0x00000] in <filename unknown>:0 
  at GooglePlayServices.PlayServicesResolver.<ScheduleAutoResolve>m__0 () [0x00000] in <filename unknown>:0 
  at Google.RunOnMainThread+ScheduledJob.PollUntilExecutionTime () [0x00000] in <filename unknown>:0 
  at Google.RunOnMainThread.ExecutePollingJobs () [0x00000] in <filename unknown>:0 
UnityEngine.Debug:LogError(Object)
Google.RunOnMainThread:ExecutePollingJobs()
Google.RunOnMainThread:ExecuteAll()
Google.RunOnMainThread:Run(Action, Boolean)
Google.RunOnMainThread:AddOnUpdateCallback(CallbackFunction)
Google.RunOnMainThread:add_OnUpdate(CallbackFunction)
GooglePlayServices.PlayServicesResolver:.cctor()
Firebase.Editor.GenerateXmlFromGoogleServicesJson:.cctor()
Firebase.Editor.GenerateXmlFromGoogleServicesJson:.cctor()
UnityEditor.AssetDatabase:Refresh()
GooglePlayGames.Editor.GPGSUpgrader:.cctor() (at Assets/GooglePlayGames/Editor/GPGSUpgrader.cs:114)
UnityEditor.EditorAssemblies:ProcessInitializeOnLoadAttributes()
```

Looks like this is being triggered during the Android build process so *something* is locking the temporary xml file that is being updated.  Smells like a race condition between the auto-resolver and manual resolution.

**Comment by stewartmiles** [comment]
Ok I think what is going on here is due to 
https://github.com/googlesamples/unity-jar-resolver/blob/8db486643e878309fc6cae6fe9bd58b4c769ad86/source/PlayServicesResolver/src/ResolverVer1_1.cs#L261

the stream is never explicitly closed which means that it will close when it gets garbage collected which could happen at any time.  I'll change this code to close and use a using statement to make sure it calls Dispose() on the object as well.

**Comment by michieal** [comment]
Just curious if this has been fixed or not. I am getting this as a build error in Unity 2018.2.2f...

**Comment by stewartmiles** [comment]
This should be fixed in our 5.2.1 release.
